Editorial: Use grammar for VLQ and Mappings and decode via SDO

szuend opened the pull request:
Draft that changes VLQ and mappings to use grammar plus "syntax-directed operations" for decoding rather than a pure algorithmic approach.

Only uploaded so we have a preview for discussion.

Preview: https://szuend.github.io/source-map/branch/mappings-grammar/#sec-mappings

nicolo-ribaudo commented on spec.emu:
```diff
@@ -275,12 +276,99 @@
+        DigitWithoutContinuationBit
+        DigitWithContinuationBit VlqDigitList
+
+      DigitWithoutContinuationBit :: one of
+        `A` `B` `C` `D` `E` `F` `G` `H` `I` `J` `K` `L` `M` `N` `O` `P` `Q` `R`
+        `S` `T` `U` `V` `W` `X` `Y` `Z` `a` `b` `c` `d` `e` `f`
+
+      DigitWithContinuationBit :: one of
```
Very nitpicky, but I would find it easier to read if there was a bigger difference between the two names. Wdyt about `TerminalDigit` and `ContinuationDigit`?

szuend replied:
I like it! Done.

nicolo-ribaudo commented:
I attempted to re-express the DecodeBase64VLQ operation without using an accumulator, since the way fields are mutated in the accumulator is not super easy to keep track of (recursion + mutable state). Do you think this version would work, or am I missing something about shifting the values?

And then you use it as "let _value_ be the VLQSignedValue of _Vlq_".

**VLQSignedValue (  )**

_Vlq_ :: _VlqDigitList_
1. Let _unsigned_ be the VLQUnsignedValue of _VlqDigitList_.
2. Let _sign_ be 1.
3. If _unsigned_ modulo 2 is 1, set _sign_ to -1.
4. Return _sign_ &times; floor(_unsigned_ / 2).

**VLQUnsignedValue ( )**

_VlqDigitList_ :: _DigitWithoutContinuationBit_
1. Return DecodeBase64Digit(_DigitWithoutContinuationBit_).

_VlqDigitList_ :: _DigitWithContinuationBit_ _VlqDigitList_
1. Let _left_ be DecodeBase64Digit(_DigitWithContinuationBit_).
5. Let _right_ be the VLQUnsignedValue of _VlqDigitList_.
6. Return (_right_ - 2<sup>32</sup>) &times; 2<sup>5</sup>.

szuend commented:
Very nice! That looks much simpler. Although I think there is a bug: We need to slice the continuation bit off _left_ and not off `_right_`:

```
3. Return _right_ × 2 ** 5 + (_left_ - 32).
```

Alternatively we could define `DecodeBase64Digit` as an SDO that handles the continuation bit.

nicolo-ribaudo commented:
You are right, the `- 32` is in the wrong place :+1:

szuend commented:
With https://github.com/tc39/ecmarkup/pull/637 still in-flight, I applied Nicolos idea but via an additional nonterminal indirection.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -275,12 +276,97 @@
+        `y` `z` `0` `1` `2` `3` `4` `5` `6` `7` `8` `9` `+` `/`
+    </emu-grammar>
+
+    <emu-note>
```
```suggestion
    <emu-note type="editor">
```

Ecmarkup supports "editorial notes", to write notes on how the spec is written rather than on its meaning.

szuend replied:
Done!

nicolo-ribaudo commented on spec.emu:
```diff
@@ -275,12 +276,97 @@
+        1. Let _unsigned_ be the VLQUnsignedValue of |VlqDigitList|.
+        1. If _unsigned_ modulo 2 = 1, let _sign_ be -1.
+        1. Else, let _sign_ be 1.
+        1. Return _sign_ × floor(_unsigned_ / 2).
```
I just realized that we need to preserve step 11 of https://tc39.es/ecma426/#sec-DecodeBase64VLQ: _If value is 0 and sign is -1, return -2<sup>31</sup>._

szuend replied:
I missed the edit and added the check as it was previously. I also added back in the checks for >= 2 ** 31, both for the signed and unsigned case. PTAL if this works as I intend it to xD

nicolo-ribaudo commented on spec.emu:
```diff
@@ -275,12 +276,97 @@
+    <emu-clause id="sec-VLQUnsignedValue" type="sdo">
+      <h1>VLQUnsignedValue ( ): an non-negative integer</h1>
+      <dl class="header"></dl>
+      <emu-grammar>
+        VlqDigitList ::
+          TerminalDigit
+      </emu-grammar>
+      <emu-alg>
+        1. Return VLQUnsignedValue of |TerminalDigit|.
+      </emu-alg>
```
This is actually not needed, as the default behavior for SDOs is to apply itself on the only child: https://tc39.es/ecma262/#sec-algorithm-conventions-syntax-directed-operations (from "Unless explicitly specified otherwise...").

szuend replied:
Done.

nicolo-ribaudo reviewed (commented):
Second review pass. The VLQs look good now, and I finally reviewed the mappings definitions.

I would prefer this slightly different approach, based on the "informal" understanding that mappings contains one or more segments, one per line, separated by semicolons. A segment contains zero or more mappings, separated by commas. This avoids having to think about the "what if there are consecutive semicolons?" case, since they are just consecutive segments.

```
Mappings :
  Segment
  Segment `;` Mappings

Segment :
  MappingList?

MappingList :
  Mapping
  Mapping `,` MappingList
```

Note that the above definitions makes `"mappings": ""` valid, which matches the current spec.

_DecodeMappingsSdo_ would then become
**DecodeMappingsSdo ( ... )**

_Mappings : Segment `;` Mappings_
_MappingList : Mapping `,` MappingList_
1. For each child node _child_ of this Parse Node, do ... (same as now)

_Segment : MappingList?_
1. Set _state_.[[GeneratedLine]] to _state_.[[GeneratedLine]] + 1.
2. Set _state_.[[GeneratedColumn]] to 0.
3. If _MappingList_ is present, perform DecodeMappingsSdo of _MappingList_ with arguments ...

_Mapping : GeneratedColumn_ (same as now)
_Mapping :  GeneratedColumn OriginalSource OriginalLine OriginalColumn Name?_ (same as now)

szuend commented:
Thanks for the new grammar, I love it!

> Maybe can we also rename _Mappings_ to _SegmentList_, or _MappingsSegmentList_? It's a bit weird that "mappings" is a list of segments, and "segment" is a list of mappings, even though that's how we always referred to it 😅

IMO it would be nice to have consistent naming of the goal symbols w.r.t. to how the field is named in the source map JSON. What about renaming _Mappings_ to _SegmentList_ but then add a goal symbol _MappingsField_ : _SegmentList_?

nicolo-ribaudo commented:
> What about renaming Mappings to SegmentList but then add a goal symbol MappingsField : SegmentList?

Sounds good 👍

szuend commented:
Changed the grammar as per your suggestion. Also renamed the SDO to `DecodeMappingsField`, which makes more sense. The "Sdo" suffix was only there because I couldn't think of anything better at the time.

takikawa commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+      <emu-clause id="sec-DecodeMappingsField" type="sdo">
+        <h1>
+          DecodeMappingsField (
+            _state_: a Decode Mapping State Record,
```
Very minor: this doesn't get hyperlinked because the variant specified above is plural but this isn't

szuend replied:
Thanks! The &lt;dfn&gt; had a typo.

takikawa commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+        <emu-alg>
+          1. Let _relativeSourceIndex_ the VLQSignedValue of |Vlq|.
+          1. Set _state_.[[SourceIndex]] to _state_.[[SourceIndex]] + _relativeSourceIndex_.
+          1. If _state_.[[SourceIndex]] &lt; 0 or _state.[[SourceIndex]] ≥ the number of elements of _sources_, optionally report an error.
```
Minor typo: the second `_state_` is missing the end `_`.

szuend replied:
Fixed :)

takikawa commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+          1. Perform DecodeMappingsField of |OriginalSource| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Perform DecodeMappingsField of |OriginalLine| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Perform DecodeMappingsField of |OriginalColumn| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Let _source_ be _sources_[_state_.[[SourceIndex]]].
```
It's only in an error case, but there's a small semantic difference between the old version and this I think. If the source index is invalid, the old algorithm kept the source as `null` while this returns `undefined` in JS or out-of-bounds lookup in general. (original line/column are also kept `null` when they are negative)

szuend replied:
Yeah there is some bound checks missing. Same for `names`.

I'm not sure about `undefined` though. We are using the [List specification type](https://tc39.es/ecma262/#sec-list-and-record-specification-type) here which does not actually spell out what is returned for out-of-bounds accesses, so it's even worse since its undefined behavior.

I'll add some bound checks.

szuend replied:
I rewrote the algorithm here to do the actual validation and set the fields explicitly to *null*. PTAL.

szuend replied:
I also filed #184. It seems some implementations (e.g. Chrome) behave as I wrote the spec initially but others (e.g. Firefox) behave as the existing spec says.

jridgewell commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+        <emu-alg>
+          1. Let _relativeName_ be the VLQSignedValue of |Vlq|.
+          1. Set _state_.[[NameIndex]] to _state_.[[NameIndex]] + _relativeName_.
+          1. If _state_.[[NameIndex]] &lt; 0 or _state_[[NameIndex]] ≥ the number of elements of _sources_, optionally report an error.
```
```suggestion
          1. If _state_.[[NameIndex]] &lt; 0 or _state_.[[NameIndex]] ≥ the number of elements of _sources_, optionally report an error.
```

szuend commented on spec.emu:
```diff
@@ -276,6 +276,95 @@
+      </emu-grammar>
+      <emu-alg>
+        1. Let _value_ be the VLQUnsignedValue of |VlqDigitList|.
+        1. If _value_ is ≥ 2<sup>31</sup>, throw an error.
```
I don't think this is the case. This algorithm is defined for the grammar rule `Vlq :: VlqDigitList`, which means it is not invoked at all for `VLQSignedValue` (since that is also defined on `Vlq :: VlqDigitList`.

jridgewell replied:
`VLQSignedValue` calls `VLQUnsignedValue` as its first step

szuend replied:
Yes, you are indeed right.

`VLQSignedValue` calls `VLQUnsignedValue` as its first step but on the `|VlqDigitList|` parse node, not the `|Vlq|` parse node. This means signed values are fine, but the unsiged ones would be truncated.

szuend commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+            Mapping `,` MappingList
+        </emu-grammar>
+        <emu-alg>
+          1. For each child node _child_ of this Parse Node, do
+            1. If _child_ is an instance of a nonterminal, then
```
I adapted this from the ECMAScript spec that has a couple instances of this (e.g. [Contains](https://tc39.es/ecma262/#sec-static-semantics-contains)).

Since we only have two productions, we could spell them also out explicitly:

```html
<emu-grammar>
  SegmentList :
    Segment `;` SegmentList
</emu-grammar>
<emu-alg>
  1. Perform DecodeMappingsField of |Segment| ...
  1. Perform DecodeMappingsFIeld of |SegmentList| ...
</eum-alg>
```

And same for the `MappingList` production.

szuend commented on spec.emu:
```diff
@@ -676,134 +765,210 @@
+            MappingList?
+        </emu-grammar>
+        <emu-alg>
+          1. Set _state_.[[GeneratedLine]] to _state_.[[GeneratedLine]] + 1.
```
You are right.

I think it might be sufficient to switch the statements around. Perform `DeocdeMappingsField` first, then increment the line.

Alternatively we could also move the increment to the `SegemtnList` production. I'd find that even clearer:

```
<emu-grammar>
  SegmentList :
    Segment `;` SegmentList
</emu-grammar>
<emu-alg>
  1. Perform DecodeMappingsField on |Segment| ...
  1. Set _state_.[[GeneratedLine]] to _state_.[[GeneratedLine]] + 1.
  1. Set _state_.[[GeneratedColumn]] to 0.
  1. Perform DecodeMappingsField on |SegmentList| ...
</emu-alg>
```

@nicolo-ribaudo any preference?

szuend replied:
As Justin suggested, I renamed "Segment" to "List". I slightly prefer the approach where we increment the line in the SDO of ``Line `;` LineList``, as the semicolon is directly there.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -276,6 +276,95 @@
     </dd>
   </dl>
+
+  <emu-clause id="sec-vlq-grammar">
```
Ideally "terms and definitions" should have "simple" definitions, easily expressible as "name -> meaning", like for a human-language dictionary.

I would prefer to move this emu-clause to the top level, remove the existing _base64 VLQ_ `<dfn>` from "terms and definitions", and use it as an introductory paragraph for this new section, including the examples.

szuend replied:
Done, I renamed the top level section to "base64 VLQ".

nicolo-ribaudo commented on spec.emu:
```diff
@@ -276,6 +276,95 @@
+        `y` `z` `0` `1` `2` `3` `4` `5` `6` `7` `8` `9` `+` `/`
+    </emu-grammar>
+
+    <emu-note type="editor">
```
Don't forget to do this after rebasing.

szuend replied:
Done

nicolo-ribaudo commented on spec.emu:
```diff
@@ -276,6 +276,95 @@
+        1. Else, let _sign_ be 1.
+        1. Let _value_ be floor(_unsigned_ / 2).
+        1. If _value_ is 0 and _sign_ is -1, return -2<sup>31</sup>.
+        1. If _value_ is ≥ 2<sup>31</sup>, throw an error.
```
- When _sign_ is -1, 2<sup>31</sup> is actually a supported value (the ranges goes from -2<sup>31</sup> to 2<sup>32</sup>-1).
- Given that we know that _unsigned_ is less than 2<sup>32</sup> (otherwise VLQUnsignedValue would have thrown), we don't actually need this check. It can be an assertion.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -276,6 +276,95 @@
+        1. Else, let _sign_ be 1.
+        1. Let _value_ be floor(_unsigned_ / 2).
+        1. If _value_ is 0 and _sign_ is -1, return -2<sup>31</sup>.
+        1. If _value_ is ≥ 2<sup>31</sup>, throw an error.
```
I would maybe add a note somewhere that we need this check because we are calling VLQUnsignedValue on |VlqDigitList| rather than on |Vlq|, so we are skipping VLQUnsignedValue's validation.

Or we call VLQUnsignedValue on |Vlq|, and replace this check with an assertion.

szuend replied:
Added a note. Otherwise we'd have to turn `VLQSignedValue` into an abstract op and pass in a |Vlq| parse node.

nicolo-ribaudo reviewed (commented):
I left minor some comments, but the rest looks good to me. I'll approve once rebased.

nicolo-ribaudo commented:
Oh also, per the last call we had, lets add a "Grammar notation" section in "Notational conventions". It should encourage readers to check out https://tc39.es/ecma262/#sec-grammar-notation, mentioning explicitly that:
- the terminal symbols of our grammars are individual code points (i.e. they are more like the lexical and regexp grammars of ECMA-262, than like the syntactic grammar)
- we do not use https://tc39.es/ecma262/#sec-grammatical-parameters and https://tc39.es/ecma262/#sec-lookahead-restrictions, which keeps our grammar simpler.

szuend commented:
> Oh also, per the last call we had, lets add a "Grammar notation" section in "Notational conventions".

Done. Might need some tweaking to the wording.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -236,6 +236,15 @@
+    <h1>Grammar Notation</h1>
+    <p>This specification follows the same grammar notation convention as defined by <a href="https://tc39.es/ecma262/#sec-grammar-notation">ECMA-262 (Grammar Notation)</a>, with the following caveats:</p>
+    <ul>
+      <li>The terminal symbols of grammars in this specification are individual code points. They are more like ECMA-262 lexical and regexp grammars than the ECMA-262 syntactic grammar.</li>
+      <li>This specification does not use <a href="https://tc39.es/ecma262/#sec-grammatical-parameters">grammatical parameters</a> or <a href="https://tc39.es/ecma262/#sec-lookahead-restrictions">lookahead restrictions</a>, which keeps the grammars simpler.</li>
```
```suggestion
      <li>The terminal symbols of grammars defined in this specification are individual code points. This is similar to ECMA-262's <a href="https://tc39.es/ecma262/#sec-lexical-and-regexp-grammars>lexical grammar</a>, rather than to ECMA-262's <a href="https://tc39.es/ecma262/#sec-syntactic-grammar">syntactic grammar</a>.</li>
      <li>This specification does not use <a href="https://tc39.es/ecma262/#sec-grammatical-parameters">grammatical parameters</a> or <a href="https://tc39.es/ecma262/#sec-lookahead-restrictions">lookahead restrictions</a>, to reduce the grammar definitions complexity.</li>
```

szuend replied:
Much appreciated!
